File: TinyScript/java/src/main/java/parser/ast/IfStmt.java
package parser.ast;

import parser.util.ParseException;
import parser.util.PeekTokenIterator;

public class IfStmt extends Stmt {
    public IfStmt() {
        super(ASTNodeTypes.IF_STMT, "if");
    }

    public static ASTNode parse(PeekTokenIterator it) throws ParseException {
        return parseIF( it);
    }

    // IfStmt -> If(Expr) Block Tail
    public static ASTNode parseIF(PeekTokenIterator it) throws ParseException {
        var lexeme = it.nextMatch("if");
        it.nextMatch("(");
        var ifStmt = new IfStmt();
        ifStmt.setLexeme(lexeme);
        var expr = Expr.parse( it);
        ifStmt.addChild(expr);
        it.nextMatch(")");
        var block = Block.parse( it);
        ifStmt.addChild(block);

        var tail = parseTail(it);
        if(tail != null) {
            ifStmt.addChild(tail);
        }
        return ifStmt;

    }

    // Tail -> else {Block} | else IFStmt | ε
    public static ASTNode parseTail(PeekTokenIterator it) throws ParseException {
        if(!it.hasNext() || !it.peek().getValue().equals("else")) {
            return null;
        }
        it.nextMatch("else");
        var lookahead = it.peek();

        if(lookahead.getValue().equals("{")) {
            return Block.parse(it);
        } else if(lookahead.getValue().equals("if")){
            return parseIF(it);
        } else {
            return null;
        }

    }

    public ASTNode getExpr() {
        return this.getChild(0);
    }

    public ASTNode getBlock(){
        return this.getChild(1);
    }

    public ASTNode getElseBlock(){

        var block = this.getChild(2);
        if(block instanceof Block) {
            return block;
        }
        return null;
    }

    public ASTNode getElseIfStmt(){
        var ifStmt = this.getChild(2);
        if(ifStmt instanceof IfStmt) {
            return ifStmt;
        }
        return null;
    }


}

File: TinyScript/js/src/parser/ast/IFStmt.js
const ASTNodeTypes = require("./ASTNodeTypes")
const {Stmt} = require('./index')
class IfStmt extends Stmt {
  constructor() {
    super(ASTNodeTypes.IF_STMT, "if");
  }

  
  getExpr() {
    return this.getChild(0);
  }

  getBlock() {
    return this.getChild(1);
  }

  getElseBlock() {
    const block = this.getChild(2);
    if (block instanceof Block) {
      return block;
    }
    return null;
  }

  getElseIfStmt() {
    const ifStmt = this.getChild(2);
    if (ifStmt instanceof IfStmt) {
      return ifStmt;
    }
    return null;
  }
}
module.exports = IfStmt

const {Expr, Block} = require('./index')

// IfStmt -> if(Expr) {Block} Tail
IfStmt.parse = (it) => {
  const lexeme = it.nextMatch("if");
  it.nextMatch("(");
  const ifStmt = new IfStmt();
  ifStmt.setLexeme(lexeme);
  const expr = Expr.parse(it);
  ifStmt.addChild(expr);
  it.nextMatch(")");
  const block = Block.parse(it);
  ifStmt.addChild(block);

  const tail = IfStmt.parseTail(it);
  if (tail != null) {
    ifStmt.addChild(tail);
  }
  return ifStmt;
}

// Tail -> else {Block} | else IFStmt | ε
IfStmt.parseTail = (it) => {
  if (
    !it.hasNext() ||
    it
      .peek()
      .getValue() !== "else"
  ) {
    return null;
  }
  it.nextMatch("else");
  const lookahead = it.peek();

  if (lookahead.getValue() === "{") {
    return Block.parse(it)
  } else if (lookahead.getValue() === "if") {
    return IfStmt.parse(it)
  } else {
    return null;
  }
}

